docs/designs/master_redesign.md

```diff
@@ -0,0 +1,136 @@
+
+For training jobs, the task manager supports dynamic sharding and partitions all training data into shards. Each task contains a data shard.
+
+### Rendezvous Server
+
+The rendezvous server supports for elastic AllReduce. After detecting worker pod number change, it creates a new worker host plan so that worker pods can re-initialize Horovod.
+
+### ElasticDL Training/Evaluation Service
+All other current master functions except the above three components. ElasticDL training/evaluation service supports for ElasticDL native training. It includes evaluation service, supporting for model saving and checkpoint, etc. 
+
+
+Thus, we can launch the master in seven component combinations:
+
+1. Elastic pod manager only;
+2. Task manager only (without fault-tolerance);
+3. Elastic pod manager + Task manager;
+4. Elastic pod manager + Rendezvous server;
+5. Elastic pod manager + Rendezvous server + Task manager;
+6. Elastic pod manager + Task manager + ElasticDL training/evaluation service;
+7. Elastic pod manager + Task manager + Rendezvous server + ElasticDL training/evaluation service.
+
+
+## Implementation
+
+Current version of the master
+
+```
+Init
+- get master addr/port, job_type
+- create HorovodRendezvousServer if needed.
+- load model_inst/optimizer/custom_data_reader
+  - model_inst used for tensorflow.python.keras.callbacks
+  - optimizer to get info for go-PS.
+  - custom_data_reader to get create_data_reader_fn for create_shard()
+- init callbacks
+- get model version from checkpoint
+- create task_d (task_dispatcher)
+- deferred_callback_create_train_end_task
+- create evaluation service if needed
+- create instance manager
+- create master service
+- timeout task check thread (recover task, remove worker if timeout)
+
+Prepare
+- start evaluation service if needed.
+- start rpc server
+- rendezvous service start if needed.
+- start instance_manager
+  - start PS
+  - start workers
+
+Run:
+  loop until no more tasks, or all workers completed.
+
+```
+
+We need to add more arguments
+
+- need_pod_manager (default:True)
+- need_task_manager (default: True)
+- need_training_service (default: True)
+- task_fault_tolerance (default: True)
+- worker_command (provided by ElasticDL operator, if none, using default `python -m elasticdl.python.worker.main args`)
+
+
+```python
+Init
+ args = parse_args()
+ 
+ # create pod manager if args.need_pod_manager
+ pod_manager = create_pod_manager_if_needed(args)
+ 
+ # create task_manager if args. need_task_manager
+ task_manager = create_task_manager_if_needed(args, pod_manager)
+ 
+ # create rendezvous server if args.distribution_strategy==AllreduceStrategy
+ rendezvous_server = create_rendezvous_server_if_needed(args, pod_manager)
+ 
+ # init ElasticDL training/evaluation service if args.need_training_service
+ training_service = init_training_service_if_needed(args)
+ 
+ # create master rpc service
+ rpc_server = create_rpc_service()
+
+
+Prepare
```

Comment:
The structure is similar to the initialization:
https://github.com/sql-machine-learning/elasticdl/blob/d2015d1a68f9b28c6a412dd75a1419d2e7b1ee03/elasticdl/python/master/servicer.py#L32-L40.

We should take the master and `MasterServicer` into consideration to design the master.

Reply:
We will leave master rpc service into another issue.